src/factory.ts

```diff
@@ -59,33 +59,33 @@ function produce(item: Item, rate: PerMinute, factory: FactoryGraph): ContainerN
     const recipe = findRecipe(item)
     let outputs: ContainerNode[] = []
     let additionalIndustries = 0
+    let catalystTransferUnit: TransferNode | undefined
     for (const container of containers) {
         additionalIndustries = Math.ceil(
             (rate + container.egress - container.ingress) / (recipe.product.quantity / recipe.time),
         )
-        // if this container holds catalyst byproducts and already has
-        // a catalyst transfer unit, then we must check that the transfer
-        // unit output can add links to these new industries
-        let catalystCheck = true
+
+        // Get the catalyst transfer unit consuming from this container, if any
         for (const consumer of container.consumers) {
             if (isTransferNode(consumer) && isCatalyst(consumer.item)) {
-                if (consumer.output === undefined) {
-                    console.log(consumer)
-                    throw new Error("Transfer unit has no output?")
-                }
-                if (!consumer.output.canAddOutgoingLinks(additionalIndustries)) {
-                    catalystCheck = false
-                }
+                catalystTransferUnit = consumer
             }
         }
+
+        // Check that we can add incoming and outgoing link to this container,
+        // and that we can add outgoing links to the catalyst transfer unit output
+        // if any
         if (
             container.canAddIncomingLinks(additionalIndustries) &&
             container.canAddOutgoingLinks(1) &&
-            catalystCheck
+            (catalystTransferUnit === undefined ||
+                (catalystTransferUnit.output !== undefined &&
+                    catalystTransferUnit.output.canAddOutgoingLinks(additionalIndustries)))
```

Comment:
```suggestion
            (catalystTransferUnit === undefined ||
                catalystTransferUnit.output?.canAddOutgoingLinks(additionalIndustries) === true)
```
You can use the optional chaining feature to get rid of that additional check. Basically, if `catalystTransferUnit.output` is `undefined`, the entire expression will return `undefined`. `catalystTransferUnit.output?.canAddOutgoingLinks(additionalIndustries)` has return type of `boolean | undefined`.